src/Cryptol/Symbolic/SBV.hs

```diff
@@ -181,15 +181,14 @@ satProve ProverCommand {..} =
   case predArgTypes pcSchema of
     Left msg -> return (Nothing, ProverError msg)
     Right ts -> do when pcVerbose $ lPutStrLn "Simulating..."
-                   (_,v) <- doSBVEval evo $
-                              do env <- Eval.evalDecls SBV extDgs mempty
-                                 Eval.evalExpr SBV env pcExpr
                    prims <- M.getPrimMap
                    runRes <- runFn $ do
                                (args, asms) <- runWriterT (mapM tyFn ts)
-                               (_,b) <- doSBVEval evo (Eval.fromVBit <$>
-                                          foldM Eval.fromVFun v (map pure args))
-                               return (foldr addAsm b asms)
+                               (safety,b) <- doSBVEval evo $
+                                   do env <- Eval.evalDecls SBV extDgs mempty
+                                      v <- Eval.evalExpr SBV env pcExpr
+                                      Eval.fromVBit <$> foldM Eval.fromVFun v (map pure args)
+                               return (foldr addAsm (SBV.svAnd safety b) asms)
```

Comment:
This same bit of code is used for both `:sat` and `:prove`, right? It's not entirely obvious that this does what I expect it should do in both cases. First of all, this function `satProve` is just too damn big, and also there seems to be a lack of documentation in this file. (I don't see any haddocks anywhere except for the module header.)

I think this should be good as soon as it is properly documented (and preferably broken up into a few smaller functions first, each documented separately).

Reply:
Yeah, both code paths use this same function.  I agree, it's quite confusing. I'll see if I can refactor it somewhat without breaking it.

Reply:
I think the commits below address this.